File: backend/app/services/registro_service.py
from sqlalchemy.orm import Session
from fastapi import HTTPException, status
from app.repositories.registro_repository import (
    get_registros_by_asamblea, 
    search_registros,
    get_registro_by_id,
    buscar_registros_para_poderes,
    buscar_registro_con_poder,
    buscar_dueno_original_poder,
    update_registro,
    verificar_control_existente,
    verificar_control_en_poderes,
    get_estadisticas_ingreso_por_hora,
    get_estadisticas_quorum_coeficiente
)
from app.repositories.asamblea_repository import get_asamblea_by_id
from uuid import UUID
from typing import Optional, Dict, Any

# Servicio para obtener un registro por ID
def get_registro_service(db: Session, registro_id: UUID):
    registro = get_registro_by_id(db, registro_id)
    if not registro:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Registro no encontrado"
        )
    return registro

# Servicio para obtener todos los registros de una asamblea
def get_registros(db: Session, asamblea_id: UUID):
    # Verificar que la asamblea existe
    asamblea = get_asamblea_by_id(db, asamblea_id)
    if not asamblea:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Asamblea no encontrada"
        )
    
    registros = get_registros_by_asamblea(db=db, asamblea_id=asamblea_id)
    
    return registros

# Servicio para buscar registros
def search_registros_service(
    db: Session,
    asamblea_id: UUID,
    cedula: Optional[str] = None,
    numero_torre: Optional[str] = None,
    numero_apartamento: Optional[str] = None,
    numero_control: Optional[str] = None
):
    # Verificar que la asamblea existe
    asamblea = get_asamblea_by_id(db, asamblea_id)
    if not asamblea:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Asamblea no encontrada"
        )
    
    registros = search_registros(
        db=db,
        asamblea_id=asamblea_id,
        cedula=cedula,
        numero_torre=numero_torre,
        numero_apartamento=numero_apartamento,
        numero_control=numero_control
    )
    
    return registros

# Servicio para buscar registros para autocompletado de poderes
def buscar_registros_para_poderes_service(
    db: Session,
    asamblea_id: UUID,
    torre: Optional[str] = None,
    apartamento: Optional[str] = None,
    numero_control: Optional[str] = None,
    limit: int = 10
):
    # Verificar que la asamblea existe
    asamblea = get_asamblea_by_id(db, asamblea_id)
    if not asamblea:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Asamblea no encontrada"
        )
    
    registros = buscar_registros_para_poderes(
        db=db,
        asamblea_id=asamblea_id,
        torre=torre,
        apartamento=apartamento,
        numero_control=numero_control,
        limit=limit
    )
    
    return registros

# Servicio para buscar registro que tiene un poder específico
def buscar_registro_con_poder_service(
    db: Session,
    asamblea_id: UUID,
    torre: str,
    apartamento: str,
    numero_control: str
):
    # Verificar que la asamblea existe
    asamblea = get_asamblea_by_id(db, asamblea_id)
    if not asamblea:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Asamblea no encontrada"
        )
    
    registro = buscar_registro_con_poder(
        db=db,
        asamblea_id=asamblea_id,
        torre=torre,
        apartamento=apartamento,
        numero_control=numero_control
    )
    
    return registro

# Servicio para actualizar registro
def update_registro_service(
    db: Session,
    registro_id: UUID,
    gestion_poderes: Optional[Dict[str, Any]] = None,
    actividad_ingreso: Optional[Dict[str, Any]] = None,
    numero_control: Optional[str] = None,
    set_numero_control_none: bool = False
):
    registro = get_registro_by_id(db, registro_id)
    if not registro:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Registro no encontrado"
        )
    
    registro_actualizado = update_registro(
        db=db,
        registro_id=registro_id,
        gestion_poderes=gestion_poderes,
        actividad_ingreso=actividad_ingreso,
        numero_control=numero_control,
        _numero_control_set_none=set_numero_control_none
    )
    
    return registro_actualizado

# Servicio para transferir un poder de un registro a otro
def transferir_poder_service(
    db: Session,
    asamblea_id: UUID,
    registro_destino_id: UUID,
    torre: str,
    apartamento: str,
    numero_control: str
):
    """
    Transfiere un poder de un registro a otro.
    1. Busca el registro que tiene el poder
    2. Remueve el poder de ese registro
    3. Agrega el poder al registro destino
    """
    # Verificar que la asamblea existe
    asamblea = get_asamblea_by_id(db, asamblea_id)
    if not asamblea:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Asamblea no encontrada"
        )
    
    # Verificar que el registro destino existe
    registro_destino = get_registro_by_id(db, registro_destino_id)
    if not registro_destino:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Registro destino no encontrado"
        )
    
    # Buscar el registro que tiene el poder
    registro_origen = buscar_registro_con_poder(
        db=db,
        asamblea_id=asamblea_id,
        torre=torre,
        apartamento=apartamento,
        numero_control=numero_control
    )
    
    if not registro_origen:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No se encontró un registro con el poder especificado"
        )
    
    # Si el poder ya está en el registro destino, no hacer nada
    if registro_origen.id == registro_destino_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="El poder ya pertenece a este registro"
        )
    
    # Remover el poder del registro origen
    poderes_origen = registro_origen.gestion_poderes or {}
    nuevo_poderes_origen = {}
    poder_removido_era_poder_1 = False
    poder_removido_key = None
    
    if isinstance(poderes_origen, dict):
        for poder_key, poder_data in poderes_origen.items():
            if isinstance(poder_data, dict):
                poder_torre = poder_data.get("torre") or poder_data.get("numero_torre") or ""
                poder_apartamento = poder_data.get("apartamento") or poder_data.get("numero_apartamento") or ""
                poder_control = poder_data.get("numero_control") or ""
                
                # Verificar si es el poder que estamos buscando
                if (poder_torre.strip().lower() == torre.strip().lower() and
                    poder_apartamento.strip().lower() == apartamento.strip().lower() and
                    poder_control.strip().lower() == numero_control.strip().lower()):
                    poder_removido_key = poder_key
                    poder_removido_era_poder_1 = (poder_key == "poder_1")
                else:
                    # Si no es el poder que estamos buscando, mantenerlo
                    nuevo_poderes_origen[poder_key] = poder_data
    
    # Si el poder removido era poder_1, dejarlo vacío
    if poder_removido_era_poder_1:
        nuevo_poderes_origen["poder_1"] = {
            "torre": "",
            "apartamento": "",
            "numero_control": ""
        }
        # Reorganizar los demás poderes para mantener la secuencia
        poderes_reorganizados = {"poder_1": nuevo_poderes_origen["poder_1"]}
        otros_poderes = {k: v for k, v in nuevo_poderes_origen.items() if k != "poder_1"}
        num = 2
        for key in sorted(otros_poderes.keys(), key=lambda x: int(x.replace("poder_", "")) if x.replace("poder_", "").isdigit() else 999):
            poderes_reorganizados[f"poder_{num}"] = otros_poderes[key]
            num += 1
        nuevo_poderes_origen = poderes_reorganizados
    elif nuevo_poderes_origen:
        # Reorganizar los poderes para mantener la secuencia (poder_1, poder_2, etc.)
        poderes_reorganizados = {}
        num = 1
        for key in sorted(nuevo_poderes_origen.keys(), key=lambda x: int(x.replace("poder_", "")) if x.replace("poder_", "").isdigit() else 999):
            poderes_reorganizados[f"poder_{num}"] = nuevo_poderes_origen[key]
            num += 1
        nuevo_poderes_origen = poderes_reorganizados
    
    # Agregar el poder al registro destino
    poderes_destino = registro_destino.gestion_poderes or {}
    if not isinstance(poderes_destino, dict):
        poderes_destino = {}
    
    # Encontrar el siguiente número de poder
    max_num = 0
    for key in poderes_destino.keys():
        if key.startswith("poder_"):
            try:
                num = int(key.replace("poder_", ""))
                max_num = max(max_num, num)
            except:
                pass
    
    nuevo_poder = {
        "torre": torre,
        "apartamento": apartamento,
        "numero_control": numero_control
    }
    
    poderes_destino[f"poder_{max_num + 1}"] = nuevo_poder
    
    # Reorganizar los poderes del destino para mantener la secuencia
    poderes_destino_reorganizados = {}
    num = 1
    for key in sorted(poderes_destino.keys(), key=lambda x: int(x.replace("poder_", "")) if x.replace("poder_", "").isdigit() else 999):
        poderes_destino_reorganizados[f"poder_{num}"] = poderes_destino[key]
        num += 1
    
    # Si no quedan poderes en el origen, crear poder_1 vacío
    if not nuevo_poderes_origen:
        nuevo_poderes_origen = {"poder_1": {"torre": "", "apartamento": "", "numero_control": ""}}
    
    # Actualizar ambos registros
    update_registro(
        db=db,
        registro_id=registro_origen.id,
        gestion_poderes=nuevo_poderes_origen
    )
    
    update_registro(
        db=db,
        registro_id=registro_destino_id,
        gestion_poderes=poderes_destino_reorganizados
    )
    
    return {
        "registro_origen": registro_origen,
        "registro_destino": registro_destino
    }

# Servicio para devolver un poder a su dueño original
def devolver_poder_service(
    db: Session,
    asamblea_id: UUID,
    registro_actual_id: UUID,
    torre: str,
    apartamento: str,
    numero_control: str
):
    """
    Devuelve un poder a su dueño original.
    1. Busca el dueño original del poder (quien tiene esos datos como su poder_1)
    2. Remueve el poder del registro actual
    3. Devuelve el poder al dueño original
    """
    # Verificar que la asamblea existe
    asamblea = get_asamblea_by_id(db, asamblea_id)
    if not asamblea:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Asamblea no encontrada"
        )
    
    # Verificar que el registro actual existe
    registro_actual = get_registro_by_id(db, registro_actual_id)
    if not registro_actual:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Registro actual no encontrado"
        )
    
    # Buscar el dueño original del poder
    dueno_original = buscar_dueno_original_poder(
        db=db,
        asamblea_id=asamblea_id,
        torre=torre,
        apartamento=apartamento,
        numero_control=numero_control
    )
    
    if not dueno_original:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="No se encontró el dueño original del poder"
        )
    
    # Si el poder ya está en el dueño original, no hacer nada
    if dueno_original.id == registro_actual_id:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="El poder ya pertenece a su dueño original"
        )
    
    # Remover el poder del registro actual
    poderes_actuales = registro_actual.gestion_poderes or {}
    nuevo_poderes_actuales = {}
    
    if isinstance(poderes_actuales, dict):
        for poder_key, poder_data in poderes_actuales.items():
            if isinstance(poder_data, dict):
                poder_torre = poder_data.get("torre") or poder_data.get("numero_torre") or ""
                poder_apartamento = poder_data.get("apartamento") or poder_data.get("numero_apartamento") or ""
                poder_control = poder_data.get("numero_control") or ""
                
                # Si no es el poder que estamos devolviendo, mantenerlo
                if not (poder_torre.strip().lower() == torre.strip().lower() and
                        poder_apartamento.strip().lower() == apartamento.strip().lower() and
                        poder_control.strip().lower() == numero_control.strip().lower()):
                    nuevo_poderes_actuales[poder_key] = poder_data
    
    # Reorganizar los poderes del registro actual
    if nuevo_poderes_actuales:
        poderes_reorganizados = {}
        num = 1
        for key in sorted(nuevo_poderes_actuales.keys(), key=lambda x: int(x.replace("poder_", "")) if x.replace("poder_", "").isdigit() else 999):
            poderes_reorganizados[f"poder_{num}"] = nuevo_poderes_actuales[key]
            num += 1
        nuevo_poderes_actuales = poderes_reorganizados
    else:
        # Si no quedan poderes, dejar poder_1 vacío
        nuevo_poderes_actuales = {"poder_1": {"torre": "", "apartamento": "", "numero_control": ""}}
    
    # Agregar el poder al dueño original
    poderes_dueno = dueno_original.gestion_poderes or {}
    if not isinstance(poderes_dueno, dict):
        poderes_dueno = {}
    
    # Verificar si el poder_1 del dueño original está vacío
    poder_1_vacio = False
    if "poder_1" in poderes_dueno:
        poder_1 = poderes_dueno["poder_1"]
        if isinstance(poder_1, dict):
            poder_1_torre = (poder_1.get("torre") or poder_1.get("numero_torre") or "").strip()
            poder_1_apto = (poder_1.get("apartamento") or poder_1.get("numero_apartamento") or "").strip()
            poder_1_control = (poder_1.get("numero_control") or "").strip()
            poder_1_vacio = not poder_1_torre and not poder_1_apto and not poder_1_control
    
    if poder_1_vacio:
        # Si poder_1 está vacío, restaurarlo ahí
        poderes_dueno["poder_1"] = {
            "torre": torre,
            "apartamento": apartamento,
            "numero_control": numero_control
        }
    else:
        # Si poder_1 no está vacío, agregar como siguiente poder
        max_num = 0
        for key in poderes_dueno.keys():
            if key.startswith("poder_"):
                try:
                    num = int(key.replace("poder_", ""))
                    max_num = max(max_num, num)
                except:
                    pass
        
        poderes_dueno[f"poder_{max_num + 1}"] = {
            "torre": torre,
            "apartamento": apartamento,
            "numero_control": numero_control
        }
    
    # Reorganizar los poderes del dueño original
    poderes_dueno_reorganizados = {}
    num = 1
    for key in sorted(poderes_dueno.keys(), key=lambda x: int(x.replace("poder_", "")) if x.replace("poder_", "").isdigit() else 999):
        poderes_dueno_reorganizados[f"poder_{num}"] = poderes_dueno[key]
        num += 1
    
    # Actualizar ambos registros
    registro_actual_actualizado = update_registro(
        db=db,
        registro_id=registro_actual_id,
        gestion_poderes=nuevo_poderes_actuales
    )
    
    dueno_original_actualizado = update_registro(
        db=db,
        registro_id=dueno_original.id,
        gestion_poderes=poderes_dueno_reorganizados
    )
    
    return {
        "registro_actual": registro_actual_actualizado,
        "dueno_original": dueno_original_actualizado
    }

# Servicio para verificar si un número de control ya está asignado
def verificar_control_existente_service(
    db: Session,
    asamblea_id: UUID,
    numero_control: str,
    registro_id_excluir: Optional[UUID] = None
):
    """
    Verifica si un número de control ya está asignado a otro registro.
    """
    registro = verificar_control_existente(
        db=db,
        asamblea_id=asamblea_id,
        numero_control=numero_control,
        registro_id_excluir=registro_id_excluir
    )
    
    return registro

# Servicio para verificar si un número de control existe en algún poder
def verificar_control_en_poderes_service(
    db: Session,
    asamblea_id: UUID,
    numero_control: str,
    registro_id_excluir: Optional[UUID] = None
):
    """
    Verifica si un número de control ya existe en algún poder de gestion_poderes.
    """
    registro = verificar_control_en_poderes(
        db=db,
        asamblea_id=asamblea_id,
        numero_control=numero_control,
        registro_id_excluir=registro_id_excluir
    )
    
    return registro

# Servicio para obtener estadísticas de ingreso por hora
def get_estadisticas_ingreso_por_hora_service(db: Session, asamblea_id: UUID):
    """
    Obtiene las estadísticas de ingreso agrupadas por hora.
    Solo cuenta actividades de tipo 'ingreso' o 'reingreso'.
    """
    # Verificar que la asamblea existe
    asamblea = get_asamblea_by_id(db, asamblea_id)
    if not asamblea:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Asamblea no encontrada"
        )
    
    estadisticas = get_estadisticas_ingreso_por_hora(db=db, asamblea_id=asamblea_id)
    
    return estadisticas

# Servicio para obtener estadísticas de quorum y coeficiente presente
def get_estadisticas_quorum_coeficiente_service(db: Session, asamblea_id: UUID):
    """
    Obtiene las estadísticas de quorum y coeficiente presente.
    """
    # Verificar que la asamblea existe
    asamblea = get_asamblea_by_id(db, asamblea_id)
    if not asamblea:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Asamblea no encontrada"
        )
    
    estadisticas = get_estadisticas_quorum_coeficiente(db=db, asamblea_id=asamblea_id)
    
    return estadisticas

File: frontend/lib/services/registroService.ts
import { apiFetch } from "@/lib/api";

export interface Registro {
  id: string;
  asamblea_id: string;
  cedula: string;
  nombre: string;
  telefono: string | null;
  correo: string | null;
  numero_torre: string | null;
  numero_apartamento: string | null;
  numero_control: string | null;
  coeficiente: number | null;
  actividad_ingreso: Record<string, any> | null;
  gestion_poderes: Record<string, any> | null;
  created_at: string;
  updated_at: string;
}

export interface SearchParams {
  cedula?: string;
  numero_torre?: string;
  numero_apartamento?: string;
  numero_control?: string;
}

export interface PoderSearchParams {
  torre?: string;
  apartamento?: string;
  numero_control?: string;
}

export interface TransferirPoderParams {
  registro_destino_id: string;
  torre: string;
  apartamento: string;
  numero_control: string;
}

export interface DevolverPoderParams {
  registro_actual_id: string;
  torre: string;
  apartamento: string;
  numero_control: string;
}

/**
 * Obtiene un registro por ID
 * 
 * @param registroId - ID del registro
 * @returns Registro
 */
export async function getRegistro(registroId: string): Promise<Registro> {
  const endpoint = `/registros/${registroId}`;

  try {
    const response = await apiFetch(endpoint, {
      method: "GET",
    });

    if (response.ok) {
      const data: Registro = await response.json();
      return data;
    }

    throw new Error(`Error al obtener registro: ${response.status}`);
  } catch (error) {
    console.error("Error al obtener registro:", error);
    throw error;
  }
}

/**
 * Obtiene todos los registros de una asamblea
 * 
 * @param asambleaId - ID de la asamblea
 * @returns Lista de registros
 */
export async function getRegistros(asambleaId: string): Promise<Registro[]> {
  const endpoint = `/registros/asamblea/${asambleaId}`;

  try {
    const response = await apiFetch(endpoint, {
      method: "GET",
    });

    if (response.ok) {
      const data: Registro[] = await response.json();
      return data;
    }

    let message = `Error al obtener registros: ${response.status}`;
    try {
      const body = await response.json();
      if (body.detail && typeof body.detail === "string") {
        message = body.detail;
      }
    } catch {
      // ignore
    }
    throw new Error(message);
  } catch (error) {
    console.error("Error al obtener registros:", error);
    throw error;
  }
}

/**
 * Busca registros de una asamblea por criterios
 * 
 * @param asambleaId - ID de la asamblea
 * @param params - Parámetros de búsqueda
 * @returns Lista de registros encontrados
 */
export async function searchRegistros(
  asambleaId: string,
  params: SearchParams
): Promise<Registro[]> {
  const searchParams = new URLSearchParams();
  
  if (params.cedula) searchParams.append("cedula", params.cedula);
  if (params.numero_torre) searchParams.append("numero_torre", params.numero_torre);
  if (params.numero_apartamento) searchParams.append("numero_apartamento", params.numero_apartamento);
  if (params.numero_control) searchParams.append("numero_control", params.numero_control);

  const endpoint = `/registros/asamblea/${asambleaId}/buscar?${searchParams.toString()}`;

  try {
    const response = await apiFetch(endpoint, {
      method: "GET",
    });

    if (response.ok) {
      const data: Registro[] = await response.json();
      return data;
    }

    throw new Error(`Error al buscar registros: ${response.status}`);
  } catch (error) {
    console.error("Error al buscar registros:", error);
    throw error;
  }
}

/**
 * Busca registros para autocompletado de poderes
 * 
 * @param asambleaId - ID de la asamblea
 * @param params - Parámetros de búsqueda (torre, apartamento, numero_control)
 * @param limit - Límite de resultados (default: 10)
 * @returns Lista de registros encontrados
 */
export async function buscarRegistrosParaPoderes(
  asambleaId: string,
  params: PoderSearchParams,
  limit: number = 10
): Promise<Registro[]> {
  const searchParams = new URLSearchParams();
  
  if (params.torre) searchParams.append("torre", params.torre);
  if (params.apartamento) searchParams.append("apartamento", params.apartamento);
  if (params.numero_control) searchParams.append("numero_control", params.numero_control);
  searchParams.append("limit", limit.toString());

  const endpoint = `/registros/asamblea/${asambleaId}/poderes/buscar?${searchParams.toString()}`;

  try {
    const response = await apiFetch(endpoint, {
      method: "GET",
    });

    if (response.ok) {
      const data: Registro[] = await response.json();
      return data;
    }

    throw new Error(`Error al buscar registros para poderes: ${response.status}`);
  } catch (error) {
    console.error("Error al buscar registros para poderes:", error);
    throw error;
  }
}

/**
 * Verifica si un poder existe y retorna el registro que lo tiene
 * 
 * @param asambleaId - ID de la asamblea
 * @param torre - Torre/Bloque
 * @param apartamento - Apto/Casa
 * @param numero_control - Número de control
 * @returns Registro que tiene el poder, o null si no existe
 */
export async function verificarPoder(
  asambleaId: string,
  torre: string,
  apartamento: string,
  numero_control: string
): Promise<Registro | null> {
  const searchParams = new URLSearchParams();
  searchParams.append("torre", torre);
  searchParams.append("apartamento", apartamento);
  searchParams.append("numero_control", numero_control);

  const endpoint = `/registros/asamblea/${asambleaId}/poderes/verificar?${searchParams.toString()}`;

  try {
    const response = await apiFetch(endpoint, {
      method: "GET",
    });

    if (response.ok) {
      const data: Registro = await response.json();
      return data;
    }

    if (response.status === 404) {
      return null;
    }

    throw new Error(`Error al verificar poder: ${response.status}`);
  } catch (error) {
    console.error("Error al verificar poder:", error);
    throw error;
  }
}

/** Campos que se pueden actualizar en un registro */
export interface RegistroUpdatePayload {
  cedula?: string;
  nombre?: string;
  telefono?: string | null;
  correo?: string | null;
  numero_torre?: string | null;
  numero_apartamento?: string | null;
  numero_control?: string | null;
  gestion_poderes?: Record<string, any> | null;
  actividad_ingreso?: Record<string, any> | null;
}

/**
 * Actualiza un registro
 *
 * @param registroId - ID del registro
 * @param payload - Campos a actualizar (solo se envían los definidos)
 * @returns Registro actualizado
 */
export async function actualizarRegistro(
  registroId: string,
  payload: RegistroUpdatePayload
): Promise<Registro> {
  const endpoint = `/registros/${registroId}`;

  const body: Record<string, unknown> = {};
  if (payload.cedula !== undefined) body.cedula = payload.cedula;
  if (payload.nombre !== undefined) body.nombre = payload.nombre;
  if (payload.telefono !== undefined) body.telefono = payload.telefono;
  if (payload.correo !== undefined) body.correo = payload.correo;
  if (payload.numero_torre !== undefined) body.numero_torre = payload.numero_torre;
  if (payload.numero_apartamento !== undefined) body.numero_apartamento = payload.numero_apartamento;
  if (payload.numero_control !== undefined) body.numero_control = payload.numero_control;
  if (payload.gestion_poderes !== undefined) body.gestion_poderes = payload.gestion_poderes;
  if (payload.actividad_ingreso !== undefined) body.actividad_ingreso = payload.actividad_ingreso;

  try {
    const response = await apiFetch(endpoint, {
      method: "PUT",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });

    if (response.ok) {
      const data: Registro = await response.json();
      return data;
    }

    throw new Error(`Error al actualizar registro: ${response.status}`);
  } catch (error) {
    console.error("Error al actualizar registro:", error);
    throw error;
  }
}

/**
 * Transfiere un poder de un registro a otro
 * 
 * @param asambleaId - ID de la asamblea
 * @param params - Parámetros de transferencia
 * @returns Resultado de la transferencia
 */
export async function transferirPoder(
  asambleaId: string,
  params: TransferirPoderParams
): Promise<{ registro_origen: Registro; registro_destino: Registro; message: string }> {
  const endpoint = `/registros/asamblea/${asambleaId}/poderes/transferir`;

  try {
    const response = await apiFetch(endpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(params),
    });

    if (response.ok) {
      const data = await response.json();
      return data;
    }

    throw new Error(`Error al transferir poder: ${response.status}`);
  } catch (error) {
    console.error("Error al transferir poder:", error);
    throw error;
  }
}

/**
 * Obtiene el coeficiente del dueño original de un poder
 * 
 * @param asambleaId - ID de la asamblea
 * @param torre - Torre/Bloque
 * @param apartamento - Apto/Casa
 * @param numero_control - Número de control
 * @returns Coeficiente del dueño original o null
 */
export async function obtenerCoeficientePoder(
  asambleaId: string,
  torre: string,
  apartamento: string,
  numero_control: string
): Promise<number | null> {
  const searchParams = new URLSearchParams();
  searchParams.append("torre", torre);
  searchParams.append("apartamento", apartamento);
  searchParams.append("numero_control", numero_control);

  const endpoint = `/registros/asamblea/${asambleaId}/poderes/coeficiente?${searchParams.toString()}`;

  try {
    const response = await apiFetch(endpoint, {
      method: "GET",
    });

    if (response.ok) {
      const data = await response.json();
      return data.encontrado ? (data.coeficiente || null) : null;
    }

    if (response.status === 404) {
      return null;
    }

    throw new Error(`Error al obtener coeficiente: ${response.status}`);
  } catch (error) {
    console.error("Error al obtener coeficiente del poder:", error);
    return null;
  }
}

/**
 * Verifica si un número de control ya está asignado a otro registro
 * 
 * @param asambleaId - ID de la asamblea
 * @param numero_control - Número de control a verificar
 * @param registroIdExcluir - ID del registro a excluir de la verificación (opcional)
 * @returns true si el control ya está asignado, false si no
 */
export async function verificarControlExistente(
  asambleaId: string,
  numero_control: string,
  registroIdExcluir?: string
): Promise<{ existe: boolean; registro: Registro | null }> {
  const searchParams = new URLSearchParams();
  searchParams.append("numero_control", numero_control);
  if (registroIdExcluir) {
    searchParams.append("registro_id_excluir", registroIdExcluir);
  }

  const endpoint = `/registros/asamblea/${asambleaId}/control/verificar?${searchParams.toString()}`;

  try {
    const response = await apiFetch(endpoint, {
      method: "GET",
    });

    if (response.ok) {
      const data = await response.json();
      return data;
    }

    throw new Error(`Error al verificar control: ${response.status}`);
  } catch (error) {
    console.error("Error al verificar control:", error);
    throw error;
  }
}

/**
 * Devuelve un poder a su dueño original
 * 
 * @param asambleaId - ID de la asamblea
 * @param params - Parámetros de devolución
 * @returns Resultado de la devolución
 */
export async function devolverPoder(
  asambleaId: string,
  params: DevolverPoderParams
): Promise<{ registro_actual: Registro; dueno_original: Registro; message: string }> {
  const endpoint = `/registros/asamblea/${asambleaId}/poderes/devolver`;

  try {
    const response = await apiFetch(endpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(params),
    });

    if (response.ok) {
      const data = await response.json();
      return data;
    }

    throw new Error(`Error al devolver poder: ${response.status}`);
  } catch (error) {
    console.error("Error al devolver poder:", error);
    throw error;
  }
}
